File: 0503-next-greater-element-ii/0503-next-greater-element-ii.py
# Similar to daily Temperatures question but it since its circular we need to iterate over it twice to find next greater number
# 
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        n = len(nums)
        mystack = []
        res = [-1] * n
        
        
        for i in range(2*n):
            while mystack and nums[mystack[-1]] < nums[i%n]:
                temp = mystack.pop()
                next_greater = nums[i%n]
                res[temp] = next_greater
            
            if res[i%n] == -1:
                mystack.append(i%n)
            
        return res

File: 0503-next-greater-element-ii/0503-next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        Stack<Integer> mystack = new Stack<>();
        int[] res = new int[n];
        Arrays.fill(res,-1);
        for (int i = 0; i< 2*n; i++){
            while(!mystack.isEmpty() && nums[mystack.peek()] < nums[i%n]){
                int prev_index = mystack.pop();
                res[prev_index] = nums[i%n];
            }
            if (i < n){
                mystack.push(i);
            }
            // else{
            //     if (mystack.peek() == i%n){
            //         break;
            //     }
            // }
            
        }
        return res;
    }
}